### ISSUE TITLE
metadata caching does not work as expected - repeatedly getting lost in arc and l2arc with primary|secondarycache=metadata

### ISSUE THREAD

**Issue by devZer0** [openzfs/zfs#12028]
Distribution Name	| debian
Distribution Version	|  10.9 (buster)
Linux Kernel	| 5.4.106-1
Architecture	| x86_64
ZFS Version	| 2.0.4-pve1
SPL Version	| 2.0.4-pve1

### Describe the problem you're observing
i have metadata intensive workload, so i set primarycache=metadata and secondarycache=metadata because otherwise, data would evict metadata too soon.

when doing "find /dir/tree/with/1mio/dirs/files" or "rsync -av --dry-run /dir/tree/with/1mio/dirs/files /tmp"  i see arc and l2arc filling, but after a while, arc is collapsing and file metadata is not even served from l2arc, or at least only being served particularly from there.

even doing find / rsync a multiple times does not reliable fix this, i see lot's of reads being served from disk instead from arc or l2arc on ssd, so this renders using arc/l2arc for metadata caching pretty useless for me.

in arcstat, i see typical arc miss% at 5 and dm% at 4-5 - and suddenly pm% goes up from 0 to 100 and is see read i/o hitting the clasic harddisks instead being served from ram/ssd .  that's very frustrating.

system has 4gb of ram, btw. 

i'm  trying to optimize two different system (rsync backup machine and proxmox backup server) for metadata centric type of workload, because otherwise backup job runtime gets too long.   i'm aware of special vdev feature for relocating metadata, but i do not see the point why i should invest into enterprise quality, mirrored ssd when darn cheap consumer ssd should be able to act as a metadata read caching device, which is not not causing any data loss when it breaks.

**Comment by gamanakis** [comment]
I think you have a misunderstanding how L2ARC works. By default it scans a few hundred megabytes at the end of ARC and caches whatever buffers are there. Meaning if your ARC content is changing fast, then L2ARC caches only part of the ARC.

Moreover, by default prefetches are not cached in L2ARC. You can turn this on by l2arc_noprefetch=0 in zfs module parameters.

Have a look at l2arc_write_max l2arc_write_boost l2arc_headroom l2arc_noprefetch parameters in `man 5 zfs-module-parameters`.

**Comment by shodanshok** [comment]
@devZer0 have you tried setting `l2arc_noprefetch=0` ?

**Comment by devZer0** [comment]
yes. does not help

**Comment by malventano** [comment]
The larger issue here is that folks should not have to limit their arc/l2arc to metadata-only just as a workaround, as it trades one performance problem for another. One example is that with metadata-only caching, a process sequentially reading blocks of a smaller size than recordsize will effectively amplify those reads. 16KB sequential reads of a dataset with recordsize=1M results in 64x1MB reads, resulting in that read taking 64x as long as it should. Perhaps that's a separate optimization that should be made elsewhere (you should still at least buffer the last record read to handle this case), but it does show how workarounds like this should not be needed in the first place and can have unintended consequences.

Users desiring more persistent metadata should be able to specify an arc_meta_min (on systems with sufficient DRAM), and that should be enough to keep that metadata in the cache. FreeNAS pre-configures arc_meta_min as several GB (vs. openzfs default of 32MB) to try and (unsuccessfully with high data throughputs) address metadata eviction. Even without arc_meta_min or other parameter tweaks, the default behavior should have metadata take sufficient priority over data that metadata is very hard to evict. I do believe that was the intent, but given the reports spanning across multiple distros and versions, it appears this is an actual bug that needs squashing, and it has existed since at least as far back as 0.8.4.

**Comment by shodanshok** [comment]
> The larger issue here is that folks should not have to limit their arc/l2arc to metadata-only just as a workaround, as it trades one performance problem for another. One example is that with metadata-only caching, a process sequentially reading blocks of a smaller size than recordsize will effectively amplify those reads. 1MB of 16KB sequential reads on a dataset with recordsize=1M results in 64x1MB reads, resulting in that read taking 64x as long as it should. Perhaps that's a separate optimization that should be made elsewhere (you should still at least buffer the last record read even with metadata-only caching to handle this case), but it does show how workarounds like this should not be needed in the first place and can have unintended consequences.

Yeah, for this reason, when I have both a metadata-rich and sequential read heavy workloads, I generally set `primarycache=all` (to avoid read amplification) and `secondarycache=metadata` (to store as much metadata as possible).

> Users desiring more persistent metadata should be able to specify an arc_meta_min (on systems with sufficient DRAM), and that should be enough to keep that metadata in the cache. FreeNAS pre-configures arc_meta_min as several GB (vs. openzfs default of 16MB) to try and (unsuccessfully with high data throughputs) address metadata eviction. Even without arc_meta_min or other parameter tweaks, the default behavior should have metadata take sufficient priority over data that metadata is very hard to evict. I do believe that was the intent, but given the reports spanning across multiple distros and versions, it appears this is an actual bug that needs squashing, and it has existed since at least as far back as 0.8.4.

True, for this reason I am (for what it is worth) against removing the `arc_meta_min` as was recently discussed.

**Comment by nim-odoo** [comment]
Hello,

We are facing the same issue for a similar use case (rsync over 1M+ files). We can mitigate the behavior with `primarycache|secondarycache=metadata`, but a `zfs send` still fills the ARC with data! It doesn't make any sense to me: why would we find data in the ARC while we specifically set it to store metadata only? Moreover, setting `secondarycache=metadata` implies a waste of resources since our metadata doesn't need the 512 GB available in the L2ARC.

As said by other users, no matter how much we try to tune `arc_meta_min` it doesn't seem to have any effect.

Below is the example of what happens at 0:00 when the rsync process starts.

With `primarycache|secondarycache=all`, metadata (yellow) is evicted although `zfs_arc_meta_min` is set to 19993325568 (~19 GB). It happens around 0:30, when a large file (46 GB) is transferred sequentially.

![Screenshot 2021-10-12 at 08-27-45 SaaS Node - Grafana](https://user-images.githubusercontent.com/11041725/136903444-e3225a2b-c84f-44c9-b378-1d9cdec51bc0.png)

With `primarycache|secondarycache=metadata`, data (green) is filled in the ARC at 2:00 AM when `zfs send` starts:

![Screenshot 2021-10-12 at 08-27-05 SaaS Node - Grafana](https://user-images.githubusercontent.com/11041725/136903448-92a41ab7-3390-4fda-9121-7efed5686eea.png)

**Comment by grahamperrin** [comment]
https://github.com/openzfs/zfs/issues/12028#issuecomment-839599847

> … l2arc_write_max, l2arc_write_boost, l2arc_headroom, l2arc_noprefetch parameters in `man 5 zfs-module-parameters`.

I can't find that manual page: 

https://openzfs.github.io/openzfs-docs/man/5/index.html

```text
% man 5 zfs-module-parameters
No manual entry for zfs-module-parameters
% man zfs-module-parameters
No manual entry for zfs-module-parameters
% apropos zfs-module-parameters
apropos: nothing appropriate
% zfs --version
zfs-2.1.99-FreeBSD_g6c8f03232
zfs-kmod-2.1.99-FreeBSD_g6c8f03232
% uname -srKU
FreeBSD 14.0-CURRENT 1400041 1400041
% 
```

**Comment by gamanakis** [comment]
Try a "man 4 zfs", they were moved to section 4 some time ago.

**Comment by devZer0** [comment]
it's extremely frustrating to see filling arc with data leads to eviction of data structures which refer to metadata in l2arc.

for example i can see sustaining 100% l2arc hits for repeated "rsync -av --dry-run /path /tmp" and as soon as other data hits the arc, metadata reads from l2arc go down to a low rate and io from metadata reads hits spinning disk again. 

so this renders secondarycache=metadata param completely unusable.

anyhow, caching metadata in arc and l2arc is severely broken imho.

**Comment by shodanshok** [comment]
@devZer0 do you have l2arc prefetch enabled? Can you share the output of `arc_summary` (with tunables)? Do you have any other dataset with `secondarycache=all`?

**Comment by devZer0** [comment]
>do you have l2arc prefetch enabled?

it is set to 0. 

it does not help regarding "too early eviction of metadata", but indeed it helps a LOT for better and intensely filling the l2arc with evicted metadata. 

when set to 1, i see filling of l2arc with evicted metadata immediately comes to a standstill.  

so, this is indeed a good pointer and very important tuneable here.

>Can you share the output of arc_summary (with tunables)? 

yes, i will do soon after the next test runs on my system.  

>Do you have any other dataset with secondarycache=all?

no, all set to metadata

**Comment by nim-odoo** [comment]
> > do you have l2arc prefetch enabled?
> 
> it is set to 0.
> 
> it does not help regarding "too early eviction of metadata", but indeed it helps a LOT for better and intensely filling the l2arc with evicted metadata.
> 
> when set to 1, i see filling of l2arc with evicted metadata immediately comes to a standstill.
> 
> so, this is indeed a good pointer and very important tuneable here.

That's interesting. Did you also try with `l2arc_mfuonly`?

**Comment by devZer0** [comment]
no, i didn't try, but i don't see the point why MFU AND MRU metadata shouldn't go both to l2arc. i want ALL metadata evicted from arc to be sent to l2arc (and getting fed back, when requested)

from the docs:
"The default is "0" , meaning both MRU and MFU data and metadata are cached"





The
--


<code class="code-inner"><span class="line"><span class="cl">default is <span class="se">\fB</span>0<span class="se">\fR</span>, meaning both MRU and MFU data and metadata are cached</span></span></code>

**Comment by devZer0** [comment]
"Users desiring more persistent metadata should be able to specify an arc_meta_min (on systems with sufficient DRAM), and that should be enough to keep that metadata in the cache"

@malventano , unfortunately it's not a param to make metadata persistent , see https://github.com/openzfs/zfs/issues/10508#issuecomment-941064328

**Comment by unphased** [comment]
Sorry if I'm missing something basic here, but if you care to any significant degree about metadata performance, wouldn't you want use a special vdev for metadata and give it a redundant set of NVMe (or optane preferably?) disks for that purpose?

**Comment by stevecs** [comment]
metadata vdev (special vdev) has problems with mixed arrays or arrays w/ 
raidz vdevs (where you can add them but can't remove them).   But even 
then it's immaterial when you have more than enough actual ram to hold 
the entire metadata set in ram which is faster than ssd/optane by orders 
of magnitude.  In large arrays of spinning disk where the VAST majority 
of hits are metadata hits this is critical.   When metadata is in ram 
(when warmed in my case using dry run rsync) I can get 150-200K iops 
when metadata is NOT in ram io hits go down to 1 - 2 thousand, so 100x 
improvement.

Data is not as important as the data size is large enough and the array 
has enough disks with prefetching to serve any client.  However normal 
user activity is doing filesystem browsing (samba/nfs/etc).  What we 
have here is a data store that is 2000x the size of ram so data is 
always going to be pushed out, and L2 is not cost effective when you're 
talking PB size systems.

**Comment by malventano** [comment]
> Sorry if I'm missing something basic here, but if you care to any significant degree about metadata performance, wouldn't you want use a special vdev for metadata and give it a redundant set of NVMe (or even optane) disks for that purpose?

That's a workaround, but it should not be necessary if the use case has DRAM >>> metadata size. It should stay in RAM, or at least be preferred and not evicted prior to evicting other single-read data.

> @malventano , unfortunately arc_meta_min is not a param to make metadata persistent , see [#10508 (comment)](https://github.com/openzfs/zfs/issues/10508#issuecomment-941064328)

Understood, but it should.

**Comment by devZer0** [comment]
it's even worse

after a longer uptime, doing a find on my dataset will result in listing files at a rate 3k/s. that does not increase after subsequent runs. 

if the system is freshly booted, find is at 5.05k/s and subsequent runs are at 43.6k/s.  
 
root@pbs01:/backuppool/ds-pve-cluster3# find . |pv -rl >/dev/null
[5.05k/s]
root@pbs01:/backuppool/ds-pve-cluster3# find . |pv -rl >/dev/null
[43.6k/s]

so, effective metadata caching apparently gets completely broken after a while....

**Comment by devZer0** [comment]
> special vdev for metadata



> mance, wou



> Sorry if I'm missing something basic here, but if you care to any significant degree about metadata performance, wouldn't you want use a special vdev for metadata and give it a redundant set of NVMe (or even optane) disks for that purpose?

why should we split our data pool to be stored on different type of harddisk medium and  invest into expensive enterprise grade, mirrored ssd when we could use ram to prefer metadata or use darn cheap consumer ssd , which should be able to act as a metadata read caching device, which is not not causing any data loss when it breaks.

i think the existing options should get fixed instead of throwing hardware at something which had worked in unix for years

**Comment by Ryushin** [comment]
I should mention that our very large 2.2PiB pool with 232 hard drives and four JBODs enclosures, we choose to go with L2ARC instead using a special.  If our head unit goes down for some reason, we can just move four SAS3 cables to another server, bring up our pool and we can be back up shortly.  The Head Unit has 32 15.4TB E1.L SSDs in it.  Ten of them are dedicated to L2ARC for 100TiB of L2ARC, along with five 25GB mirrors for SLOG.  If this head unit goes down, and we used specials, I couldn't just move the JBODs and get back up and running.

On a nightly basis I have a cron job that crawls the entire file system to keep the metadata in ARC and L2ARC.  So the crawl looks like this:
Cold ARC and L2ARC: 65 minutes
Warm L2ARC: 43 seconds
Warm ARC: 16 seconds
We have a 81% L2ARC hit ratio.  Performance is so good using L2ARC we have no interest in limiting the system by adding a Special.  If there was an option to never evict metadata we would totally turn that on.

`
arc_summary -s l2arc

------------------------------------------------------------------------
ZFS Subsystem Report                            Fri Oct 14 07:10:28 2022
Linux 5.18.0-0.bpo.1-amd64                               2.1.5-1~bpo11+1
Machine: sb1 (x86_64)                                    2.1.5-1~bpo11+1

L2ARC status:                                                    HEALTHY
        Low memory aborts:                                             0
        Free on write:                                             75.0k
        R/W clashes:                                                 406
        Bad checksums:                                                 0
        I/O errors:                                                    0

L2ARC size (adaptive):                                          96.2 TiB
        Compressed:                                    97.7 %   94.0 TiB
        Header size:                                  < 0.1 %    9.1 GiB
        MFU allocated size:                            69.1 %   65.0 TiB
        MRU allocated size:                            30.8 %   29.0 TiB
        Prefetch allocated size:                      < 0.1 %   42.2 GiB
        Data (buffer content) allocated size:         100.0 %   94.0 TiB
        Metadata (buffer content) allocated size:     < 0.1 %    8.8 GiB

L2ARC breakdown:                                                  360.7M
        Hit ratio:                                     81.1 %     292.4M
        Miss ratio:                                    18.9 %      68.2M
        Feeds:                                                      1.7M

L2ARC writes:
        Writes sent:                                    100 %       1.4M

L2ARC evicts:
        Lock retries:                                              35.4k
        Upon reading:                                                  0
`

**Comment by malventano** [comment]
With 6M files across ~1.5PiB here (384GB DRAM, 188G for ARC), I run a find every 5 minutes to try and keep metadata hot. It normally takes 20-30s, but every so often I'll see this process jump to 10-20 minutes (cold ARC is closer to an hour). This happens with the system mostly idle - nothing beyond the occasional movement of 10-20GB of data and/or 10MB/s streaming of some other contents, and that data is only write-once / read-once when there is access.

Metadata hits every 5 minutes for weeks should be the MFU of all MFUs, and yet it's still being randomly ejected if someone happens to look at the server the wrong way :).

**Comment by shodanshok** [comment]
@Ryushin @malventano Just to be sure to understand... did you set L2ARC to only cache metadata via `secondarycache=metadata`?

**Comment by malventano** [comment]
> @Ryushin @malventano Just to be sure to understand... did you set L2ARC to only cache metadata via `secondarycache=metadata`?

There is no L2ARC in this system. I'm providing the data points here to demonstrate that the root of the issue exists at a deeper level (in the ARC). I submit that those running `secondarycache=metadata` would not need to if the default ARC/L2ARC handled metadata properly.

**Comment by Ryushin** [comment]
> @Ryushin @malventano Just to be sure to understand... did you set L2ARC to only cache metadata via `secondarycache=metadata`?

No we have not turned on secondarycache=metadata as we want it cache other content as well and we have prefetch turned on.  I run that cron job to keep the metadata in our ARC and L2ARC.  That is a bandaid to actually having an option to not evict metadata.  I don't mind if the metadata gets recycled back into ARC before being dropped, though it would be better to move that metadata in L2ARC back to the top of the L2ARC ring instead of evicting it.   If the metadata is actually needed then of course it should go back into ARC like it currently does.

**Comment by shodanshok** [comment]
I did some testing for https://github.com/openzfs/zfs/issues/14054 and I came to the following conclusions:

- metadata can be easily evicted by data because they allocate some buffer which are never promoted to the "main" MFU/MRU lists. See the following `arc_summary` after having read multiple times metadata ~100000 files via `time find /tank/test/fsmark/ -exec stat {} \+ > /dev/null`
```
[root@localhost ~]# arc_summary
ARC size (current):                                    37.5 %  735.7 MiB
        Target size (adaptive):                        27.0 %  529.8 MiB
        Min size (hard limit):                          6.2 %  122.6 MiB
        Max size (high water):                           16:1    1.9 GiB
        Most Frequently Used (MFU) cache size:         97.7 %  206.6 MiB
        Most Recently Used (MRU) cache size:            2.3 %    4.9 MiB
        Metadata cache size (hard limit):              75.0 %    1.4 GiB
        Metadata cache size (current):                 49.9 %  735.0 MiB
        Dnode cache size (hard limit):                 69.6 %    1.0 GiB
        Dnode cache size (current):                    27.7 %  284.0 MiB
```
Note how MFU+MRU is 211 MB, while metadata cache size is 735 MB with dnode cache size at 284 MB. From `arcstats` one can see that `metadata_size` really is 210 MB, but one has to add `overhead_size` at 165 MB, `dbuf_size` at 114 MB and `dnode_size` at 284 MB for a grand total of 773 MB (in line with what reported by `arc_summary` as metadata). This also means that metadata caching is, due to the various buffer, quite expensive: just compare `size` at 735 MB with `uncompressed_size` at 166 MB. Yeah, actual ARC size is *bigger* than the "uncompressed" size: this is a `compression=off` dataset, so the discrepancy can only be due to the various additional buffer consuming more memory than raw metadata size would suggest. All these additional buffers can (and should) be evicted by data, lowering metadata performance. Side note: I also tested on a `compression=lz4` dataset, and the results are the same.

- writing new data can evict metadata in the main MRU list unless one set `arc_meta_min` to a reasonable value. Note that `arc_meta_min` does not avoid ARC to cache data, but will never evict metadata if the min is not reached. This tunable only accept an absolute value, so one can not automatically set `arc_meta_min` equal to `zfs_arc_meta_limit_percent` and call it a day, but a calculation in bytes must be done. Alternatively, one can set `zfs_arc_meta_prune=0` but this means ARC will (almost) never shrink, potentially leading to out-of-memory condition.

- however, reading https://github.com/openzfs/zfs/issues/10508, `arc_meta_min` seems somewhat broken, so I am not sure how helpful would be to relying on it. On the other side, the default 16 MB value is way too small, and increasing it to a reasonable value should not pose any issue.

**Comment by adamdmoss** [comment]
If someone's eager to do some additional testing then please allow me to be mildly annoying and shill for my branch at:
https://github.com/openzfs/zfs/pull/14014

**Comment by shodanshok** [comment]
@adamdmoss first, thanks for your draft. I just tried it, below you can read my findings, comments are inline (test dataset has `compression=off`).

```
# create 100k files
[root@localhost tmp]# fs_mark -k -S 0 -D 100 -N 1000 -L 100 -n 1000 -s 0 -d /tank/test/fsmark/

# stat all them
[root@localhost tmp]# time find /tank/test/fsmark/ -exec stat {} \+ > /dev/null
real    0m36.268s
user    0m5.167s
sys     0m9.424s

# current arc stats (note metadata at over 768 MB)
ARC size (current):                                    41.7 %  817.4 MiB
        Target size (adaptive):                       100.0 %    1.9 GiB
        Min size (hard limit):                          6.2 %  122.6 MiB
        Max size (high water):                           16:1    1.9 GiB
        Most Frequently Used (MFU) cache size:         79.6 %  204.5 MiB
        Most Recently Used (MRU) cache size:           20.4 %   52.5 MiB
        Metadata cache size (hard limit):              75.0 %    1.4 GiB
        Metadata cache size (current):                 52.2 %  768.5 MiB
        Dnode cache size (hard limit):                 10.0 %  147.2 MiB
        Dnode cache size (current):                   191.4 %  281.7 MiB

# succesive find run are much faster...
[root@localhost tmp]# time find /tank/test/fsmark/ -exec stat {} \+ > /dev/null
real    0m7.956s
user    0m2.938s
sys     0m3.274s

# ...because there are no misses
[root@localhost ~]# arcstat -f time,read,miss,miss%,dmis,dm%,pmis,pm%,mmis,mm%,size,c,avail,mfu,mru,mfug,mrug 1
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  size     c  avail   mfu   mru  mfug  mrug
05:54:25   589     0      0     0    0     0    0     0    0  818M  1.9G   1.9G   587     2     0     0
05:54:26   115     0      0     0    0     0    0     0    0  817M  1.9G   1.9G   115     0     0     0
05:54:27   290     0      0     0    0     0    0     0    0  817M  1.9G   1.9G   290     0     0     0
05:54:28   109     0      0     0    0     0    0     0    0  817M  1.9G   1.9G   109     0     0     0
05:54:29   225     0      0     0    0     0    0     0    0  817M  1.9G   1.9G   225     0     0     0
05:54:30   177     0      0     0    0     0    0     0    0  817M  1.9G   1.9G   177     0     0     0
05:54:31   161     0      0     0    0     0    0     0    0  817M  1.9G   1.9G   161     0     0     0
05:54:32   174     0      0     0    0     0    0     0    0  817M  1.9G   1.9G   174     0     0     0

# read some big files
[root@localhost tmp]# dd if=/tank/test/test1.img of=/dev/null bs=2M
512+0 records in
512+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 1.79703 s, 598 MB/s
[root@localhost tmp]# dd if=/tank/test/test1.img of=/dev/null bs=2M
512+0 records in
512+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 0.7311 s, 1.5 GB/s
[root@localhost tmp]# dd if=/tank/test/test2.img of=/dev/null bs=2M
512+0 records in
512+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 1.0646 s, 1.0 GB/s
[root@localhost tmp]# dd if=/tank/test/test2.img of=/dev/null bs=2M
512+0 records in
512+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 0.545923 s, 2.0 GB/s

# current arc stats (note how metadata size is drastically reduced)
ARC size (current):                                    99.5 %    1.9 GiB
        Target size (adaptive):                       100.0 %    1.9 GiB
        Min size (hard limit):                          6.2 %  122.6 MiB
        Max size (high water):                           16:1    1.9 GiB
        Most Frequently Used (MFU) cache size:         99.9 %    1.9 GiB
        Most Recently Used (MRU) cache size:            0.1 %    1.4 MiB
        Metadata cache size (hard limit):              75.0 %    1.4 GiB
        Metadata cache size (current):                  2.8 %   41.9 MiB
        Dnode cache size (hard limit):                 10.0 %  147.2 MiB
        Dnode cache size (current):                     0.1 %  121.1 KiB

# now find is slow...
[root@localhost tmp]# time find /tank/test/fsmark/ -exec stat {} \+ > /dev/null
real    0m37.186s
user    0m5.169s
sys     0m9.707s

# ...because there are many misses 
[root@localhost ~]# arcstat -f time,read,miss,miss%,dmis,dm%,pmis,pm%,mmis,mm%,size,c,avail,mfu,mru,mfug,mrug 1
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  size     c  avail   mfu   mru  mfug  mrug
05:57:55   20K  2.6K     13  2.2K   11   450   79   722    4  1.9G  1.9G   1.4G   18K     1  2.6K     1
05:57:56   26K  2.7K     11  2.7K   11     0    0    23    1  1.9G  1.9G   1.4G   24K     0  2.7K     0
05:57:57   30K  3.3K     11  2.9K   10   438   78   454    2  1.9G  1.9G   1.4G   26K     0  3.3K     0
05:57:58   27K  3.0K     11  3.0K   11     0    0   267    2  1.9G  1.9G   1.4G   24K     0  3.0K     0
05:57:59   29K  3.3K     12  2.8K   10   438   79   440    2  1.9G  1.9G   1.4G   26K     0  3.3K     0
05:58:00   27K  3.0K     11  3.0K   11     0    0   318    2  1.9G  1.9G   1.3G   24K     0  3.0K     0
```

Lets try with `xattr=sa` (rather than default `on`):

```
# stat all files
[root@localhost tmp]# time find /tank/test/fsmark/ -exec stat {} \+ > /dev/null
real    0m10.999s
user    0m3.270s
sys     0m5.018s

# current arc stats (note the much lower metadata size)
ARC size (current):                                    12.8 %  251.6 MiB
        Target size (adaptive):                       100.0 %    1.9 GiB
        Min size (hard limit):                          6.2 %  122.6 MiB
        Max size (high water):                           16:1    1.9 GiB
        Most Frequently Used (MFU) cache size:         68.2 %   59.7 MiB
        Most Recently Used (MRU) cache size:           31.8 %   27.8 MiB
        Metadata cache size (hard limit):              75.0 %    1.4 GiB
        Metadata cache size (current):                 17.1 %  251.6 MiB
        Dnode cache size (hard limit):                 10.0 %  147.2 MiB
        Dnode cache size (current):                    63.8 %   93.9 MiB

# find is fast...
[root@localhost tmp]# time find /tank/test/fsmark/ -exec stat {} \+ > /dev/null
real    0m7.917s
user    0m2.962s
sys     0m3.356s

# ... because we have no misses
[root@localhost ~]# arcstat -f time,read,miss,miss%,dmis,dm%,pmis,pm%,mmis,mm%,size,c,avail,mfu,mru,mfug,mrug 1
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  size     c  avail   mfu   mru  mfug  mrug
06:12:48   232     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G   232     0     0     0
06:12:49   224     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G   224     0     0     0
06:12:50   221     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G   221     0     0     0
06:12:51   271     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G   271     0     0     0
06:12:52     0     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G     0     0     0     0
06:12:53   235     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G   235     0     0     0
06:12:54   244     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G   244     0     0     0

# read some big files
[root@localhost tmp]# dd if=/tank/test/test1.img of=/dev/null bs=1M
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 0.957967 s, 1.1 GB/s
[root@localhost tmp]# dd if=/tank/test/test1.img of=/dev/null bs=1M
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 0.316691 s, 3.4 GB/s
[root@localhost tmp]# dd if=/tank/test/test2.img of=/dev/null bs=1M
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 1.00167 s, 1.1 GB/s
[root@localhost tmp]# dd if=/tank/test/test2.img of=/dev/null bs=1M
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 0.212844 s, 5.0 GB/s

# find is only slightly slower...
[root@localhost tmp]# time find /tank/test/fsmark/ -exec stat {} \+ > /dev/null
real    0m8.002s
user    0m2.966s
sys     0m3.377s

# ...but only because we have much less metadata to read, still misses are present
[root@localhost ~]# arcstat -f time,read,miss,miss%,dmis,dm%,pmis,pm%,mmis,mm%,size,c,avail,mfu,mru,mfug,mrug 1
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  size     c  avail   mfu   mru  mfug  mrug
06:06:26   616   130     22     2    1   128  100   130   22  1.9G  1.9G   1.2G   485     1   130     0
06:06:27   605   129     22     1    1   128  100   129   22  1.9G  1.9G   1.2G   475     1   129     0
06:06:28     0     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G     0     0     0     0
06:06:29   605   129     22     1    1   128  100   129   22  1.9G  1.9G   1.2G   475     1   129     0
06:06:30   736   155     22     1    1   154  100   155   22  1.9G  1.9G   1.2G   580     1   155     0
06:06:31   625   131     21     1    1   130  100   131   21  1.9G  1.9G   1.2G   493     1   131     0
06:06:32     0     0      0     0    0     0    0     0    0  1.9G  1.9G   1.2G     0     0     0     0
06:06:33   631   131     21     1    1   130  100   131   21  1.9G  1.9G   1.2G   499     1   131     0
```

So, while I really like your patch because it simplifies a (partially broken?) complex code, it prevents metadata to be retained in face of significant data read/write. I think the core idea behind the current metadata/data eviction is to privilege metadata to avoid expensive small metadata reads, a behavior completely subverted by the tested patch.

Regards.

**Comment by devZer0** [comment]
>So, while I really like your patch because it simplifies a (partially broken?) complex code, 
>it prevents metadata to be retained in face of significant data read/write. 
>I think the core idea behind the current metadata/data eviction is to privilege metadata 
>to avoid expensive small metadata reads, a behavior completely subverted by the tested patch.

on my backup server, i'm still trying to get metadata be preferred/retained and i have put 

options zfs zfs_arc_meta_balance=500000
options zfs zfs_arc_max=59055800320
options zfs zfs_arc_meta_limit_percent=95
options zfs zfs_arc_dnode_limit_percent=50

in modprobe.d/zfs.conf 

system has 64gb of ram

after crawling trough the backupdir and see all metadata getting cached, after some backup activity,  i still see MASSIVE amounts of metadata reads hit the disks, i.e. even with this new code/logic,  data seems to cause eviction of metadata or dnodes or both.

this really really sucks.  

it sucks, because the problem exists for so long and so much work has been put into "enhancing" metadata/data handling - and still there seems to be no solution to prefer metadata over data in ram or at least make it "sticky" enough to survive some pressure from data reads/writes.

**Comment by devZer0** [comment]
with those settings, even with some  hourly "find /pool" on my large zpool , metadata is getting evicted

there is NO way to pin metadata in arc 😩

while true;do /usr/bin/time -f "%e" /root/crawl.sh >>crawl_times.log 2>&1 ;sleep 3600;done

47.41
43.37
48.06
51.35
49.83
65.51
63.46
59.98
54.18
47.44
154.53
69.58
67.30
3925.32
81.00
88.55
72.71
74.51
75.07

**Comment by stevecs** [comment]
Yeah, it's an ongoing issue that really sucks.  I also have an hourly cron job traversing the entire pool to keep metadata 'hot' but it's also a crap shoot.

**Comment by devZer0** [comment]
if that worked, then i would not complain loudly.  

but it only worked to some degree and then as you can see, after running successfully for a while, the crawling time again skyrockets...for no apparent reason.  that happens just out of nowhere.

**Comment by stevecs** [comment]
right, as I mentioned it is NOT a solution as it doesn't work but it's literally the *only* thing that we can do that I'm aware of for what little benefit it does provide.   This has been an issue for a very long time (where zfs_arc_meta_min is not honored at all ) and doesn't matter how much ram you have, metadata is always being pushed out of ARC in preference for just normal data).

**Comment by gmelikov** [comment]
@devZer0 maybe I've missed, but could you tell your zfs version? I'm interested if it's fresh 2.2.* or not.

**Comment by devZer0** [comment]
i'm using zfs-2.2.2-pve1

**Comment by gamanakis** [comment]
@devZer0 I think one could modify arc_evict() to skip entirely the eviction of metadata, or at least de-prioritize it. Would you be interested in trying such a patch?

**Comment by devZer0** [comment]
if there are chances for inclusion, yes

**Comment by shodanshok** [comment]
> @devZer0 I think one could modify arc_evict() to skip entirely the eviction of metadata, or at least de-prioritize it. Would you be interested in trying such a patch?

Count me interested as well. However, I think that setting `zfs_arc_meta_prune=0` *should* already avoid metadata eviction, right?

That said, this should not be needed - after all, we already have `arc_meta_min` and MFU list. Any idea on why are these not sufficient to avoid metadata eviction? Thanks.

**Comment by devZer0** [comment]
didn't zfs_arc_meta_balance patch remove zfs_arc_meta_prune and zfs_arc_meta_min ?

https://github.com/openzfs/zfs/commit/a8d83e2a24de6419dc58d2a7b8f38904985726cb

i don't see these anymore

root@pbsvm:/sys/module/zfs/parameters# find .|grep mfu
./l2arc_mfuonly

root@pbsvm:/sys/module/zfs/parameters# find .|grep arc_meta
./l2arc_meta_percent
./zfs_arc_meta_balance

**Comment by shodanshok** [comment]
You are right, `zfs_arc_meta_min` was dropped with zfs 2.2.x.
I need to give a look at this latest version, and especially to at `zfs_arc_meta_balance`

**Comment by devZer0** [comment]
not stale

**Comment by devZer0** [comment]
https://discourse.practicalzfs.com/t/keeping-metadata-in-arc/
